### PR TITLE
Customized block time in IexecHubAbstractService

### PR DESCRIPTION
**Pull request by mcornaton**

### CODE REVIEW

File: src/main/java/com/iexec/common/chain/IexecHubAbstractService.java [hunk 1]
```diff
@@ -115,6 +113,7 @@ public IexecHubAbstractService(Credentials credentials,
         this.web3jAbstractService = web3jAbstractService;
         this.iexecHubAddress = iexecHubAddress;
         this.nbBlocksToWaitPerRetry = nbBlocksToWaitPerRetry;
+        this.blockTime = blockTime;
```
Comment (james-toussaint) [inline src/main/java/com/iexec/common/chain/IexecHubAbstractService.java]: `blockTime > 0 ? blockTime: DEFAULT_BLOCK_TIME` (same applies when a DEFAULT_BLOCK_TIME appears)

Comment (mcornaton) [reply src/main/java/com/iexec/common/chain/IexecHubAbstractService.java]: Good idea, thanks!

File: src/main/java/com/iexec/common/chain/IexecHubAbstractService.java [hunk 2]
```diff
@@ -109,12 +110,37 @@ public IexecHubAbstractService(Credentials credentials,
...
+                                   Duration blockTime,
+                                   int nbBlocksToWaitPerRetry,
+                                   int maxRetries) {
         this.credentials = credentials;
         this.web3jAbstractService = web3jAbstractService;
         this.iexecHubAddress = iexecHubAddress;
         this.nbBlocksToWaitPerRetry = nbBlocksToWaitPerRetry;
-        this.blockTime = blockTime;
-        this.retryDelay = nbBlocksToWaitPerRetry * blockTime;
+        this.blockTime = blockTime.toMillis() > 0 ? blockTime : Duration.ofMillis(DEFAULT_BLOCK_TIME);
```
Comment (james-toussaint) [inline src/main/java/com/iexec/common/chain/IexecHubAbstractService.java]: What if blocktime is null?

Comment (mcornaton) [reply src/main/java/com/iexec/common/chain/IexecHubAbstractService.java]: Well... NPE. This should probably not happen, but here is a fix.

File: src/main/java/com/iexec/common/chain/IexecHubAbstractService.java [hunk 3]
```diff
@@ -111,11 +110,43 @@ public IexecHubAbstractService(Credentials credentials,
...
+                                   int nbBlocksToWaitPerRetry,
+                                   int maxRetries) {
         this.credentials = credentials;
         this.web3jAbstractService = web3jAbstractService;
         this.iexecHubAddress = iexecHubAddress;
         this.nbBlocksToWaitPerRetry = nbBlocksToWaitPerRetry;
-        this.retryDelay = nbBlocksToWaitPerRetry * blockTime;
+        if (blockTime == null || blockTime.toMillis() <= 0) {
+            log.warn("Blocktime value is incorrect, using default value [blockTime:{}, DEFAULT_BLOCK_TIME:{}]",
+                    blockTime == null ? "null" : blockTime, DEFAULT_BLOCK_TIME);
```
Comment (zguesmi) [inline src/main/java/com/iexec/common/chain/IexecHubAbstractService.java]: There is no need for this
```suggestion
                    blockTime, DEFAULT_BLOCK_TIME);
```

Comment (mcornaton) [reply src/main/java/com/iexec/common/chain/IexecHubAbstractService.java]: You're right! Updated, thanks :)

File: src/main/java/com/iexec/common/chain/IexecHubAbstractService.java [hunk 4]
```diff
@@ -111,11 +110,43 @@ public IexecHubAbstractService(Credentials credentials,
                                    int blockTime,
                                    int nbBlocksToWaitPerRetry,
                                    int maxRetries) {
+        this(
+                credentials,
+                web3jAbstractService,
+                iexecHubAddress,
+                Duration.ofMillis(blockTime > 0 ? blockTime : DEFAULT_BLOCK_TIME),
```
Comment (james-toussaint) [inline src/main/java/com/iexec/common/chain/IexecHubAbstractService.java]: I don't think we need that. If block time is not > 0, below base constructor will build duration with DEFAULT_BLOCK_TIME.
I would also add a deprecation tag on this method (comment needs to be updated too, it is not the base constructor anymore).

Comment (mcornaton) [reply src/main/java/com/iexec/common/chain/IexecHubAbstractService.java]: Right, thanks.
I was also thinking about deprecating this constructor but was unsure about that. So thanks again for pointing this out :)

File: src/main/java/com/iexec/common/chain/IexecHubAbstractService.java [hunk 5]
```diff
@@ -97,25 +96,57 @@ public IexecHubAbstractService(Credentials credentials,
...
+                                   Duration blockTime,
+                                   int nbBlocksToWaitPerRetry,
+                                   int maxRetries) {
         this.credentials = credentials;
         this.web3jAbstractService = web3jAbstractService;
         this.iexecHubAddress = iexecHubAddress;
         this.nbBlocksToWaitPerRetry = nbBlocksToWaitPerRetry;
-        this.retryDelay = nbBlocksToWaitPerRetry * blockTime;
+        if (blockTime == null || blockTime.toMillis() <= 0) {
+            log.warn("Blocktime value is incorrect, using default value [blockTime:{}, DEFAULT_BLOCK_TIME:{}]",
```
Comment (james-toussaint) [inline src/main/java/com/iexec/common/chain/IexecHubAbstractService.java]: Line should wrapped after 80

Comment (mcornaton) [reply src/main/java/com/iexec/common/chain/IexecHubAbstractService.java]: Right, that's fixed. My IDE settings were wrong :sweat_smile:

**Review by james-toussaint — APPROVED** [review]

**Review by zguesmi — APPROVED** [review]

**Pull request merged**